File: app/api/_lib/session.ts
import { randomBytes } from "node:crypto"
import { redisGet, redisSet, redisDel } from "./redis.js"

const SESSION_PREFIX = "session:"
const SESSION_TTL = 60 * 60 * 24 * 7 // 7日

export interface SessionData {
  username: string
  createdAt: number
  lastAccessedAt: number
}

export function generateToken(): string {
  return randomBytes(32).toString("hex")
}

export async function createSession(username: string): Promise<string> {
  const token = generateToken()
  const data: SessionData = {
    username,
    createdAt: Date.now(),
    lastAccessedAt: Date.now(),
  }
  await redisSet(`${SESSION_PREFIX}${token}`, data, SESSION_TTL)
  return token
}

export async function validateSession(
  token: string,
): Promise<SessionData | null> {
  if (!token) return null
  const data = await redisGet<SessionData>(`${SESSION_PREFIX}${token}`)
  if (!data) return null

  // 最終アクセス時刻を更新してセッションを延長
  data.lastAccessedAt = Date.now()
  await redisSet(`${SESSION_PREFIX}${token}`, data, SESSION_TTL)
  return data
}

export async function deleteSession(token: string): Promise<void> {
  await redisDel(`${SESSION_PREFIX}${token}`)
}

File: server/internal/domain/session/session.go
// session パッケージ: Redis を使ったセッション管理
// app/api/_lib/session.ts に相当
package session

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"time"

	redisclient "orlder-car-line/server/internal/infra/redis"
)

const (
	prefix = "session:"
	TTL    = 36 * time.Hour
)

// Data はセッションに保存するデータ。
type Data struct {
	Username       string `json:"username"`
	CreatedAt      int64  `json:"createdAt"`
	LastAccessedAt int64  `json:"lastAccessedAt"`
}

// GenerateToken はランダムな 64 文字の hex トークンを生成する。
func GenerateToken() (string, error) {
	b := make([]byte, 32)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

// Create はセッションを作成して Redis に保存し、トークンを返す。
func Create(ctx context.Context, username string) (string, error) {
	token, err := GenerateToken()
	if err != nil {
		return "", err
	}
	now := time.Now().UnixMilli()
	data := Data{
		Username:       username,
		CreatedAt:      now,
		LastAccessedAt: now,
	}
	if err := redisclient.Set(ctx, prefix+token, data, TTL); err != nil {
		return "", err
	}
	return token, nil
}

// Validate はトークンを検証し、有効なら最終アクセス時刻を更新して Data を返す。
func Validate(ctx context.Context, token string) (*Data, error) {
	if token == "" {
		return nil, nil
	}
	data, err := redisclient.Get[Data](ctx, prefix+token)
	if err != nil || data == nil {
		return nil, err
	}
	data.LastAccessedAt = time.Now().UnixMilli()
	if err := redisclient.Set(ctx, prefix+token, data, TTL); err != nil {
		return nil, err
	}
	return data, nil
}

// Delete はセッションを Redis から削除する。
func Delete(ctx context.Context, token string) error {
	return redisclient.Del(ctx, prefix+token)
}